### PR TITLE
Attempt to reduce code duplication in BCs

### PR DESCRIPTION
**Pull request by majosm**
This PR:

1. Splits the individual aspects of each BC into separate `_*BoundaryComponent` classes (e.g. `_AdiabaticBoundaryComponent`, `_NoSlipBoundaryComponent`) so that the code can be more easily reused. The fluid boundary objects now use these internally to get the respective boundary subcomponent values when assembling boundary state, etc.
2. Replaces the Mengaldo inviscid/viscous flux methods with calls to standalone flux functions that have callback inputs for boundary values. After the above changes, these were pretty much all doing the same thing and didn't need to be defined separately for each BC. 
3. Adds a `replace_fluid_state` function to eliminate some of the boilerplate code for creating modified states in the BCs.

I also made an attempt here to be a bit more precise about variable naming, specifically whether something is a Mengaldo-like boundary quantity (to be passed into a flux function) vs. an exterior face value (to be used in conjuction with an interior face value in a numerical flux). The former I've tried to label `*_bc`, the latter `*_plus`.

These changes have been applied to `AdiabaticSlipBoundary`, `IsothermalWallBoundary`, `AdiabaticNoslipWallBoundary`, and `SymmetryBoundary`. I haven't looked at the others (inflow/outflow?) yet to see if anything needs to be done to those. I also made a draft of the wall-coupling BCs after this change: [link](https://github.com/majosm/mirgecom/blob/722b0abee6e635751a530981cc57203368d5afd8/mirgecom/multiphysics/thermally_coupled_fluid_wall.py#L121-L550). (_Note: this is slightly out of date and doesn't include the `replace_fluid_state` changes._)

Depends on #678.

**Questions for the review**:
- [ ] Is the scope and purpose of the PR clear?
  - [ ] The PR should have a description.
  - [ ] The PR should have a guide if needed (e.g., an ordering).
- [ ] Is every top-level method and class documented? Are things that should be documented actually so?
- [ ] Is the interface understandable? (I.e. can someone figure out what stuff does?) Is it well-defined?
- [ ] Does the implementation do what the docstring claims?
- [ ] Is everything that is implemented covered by tests?
- [ ] Do you see any immediate risks or performance disadvantages with the design? Example: what do interface normals attach to?

### CODE REVIEW

**Comment by majosm** [comment]
@anderson2981 , @MTCam , @tulioricci I think this is ready for a look through now. Suggestion: the diff in `boundary.py` gets a little nightmarish once you get to the boundary classes, so for those it might be better to look at them side-by-side with the #678 versions instead of trying to parse it. I only modified the Mengaldo BCs: `AdiabaticSlipBoundary`, `IsothermalWallBoundary`, `AdiabaticNoslipWallBoundary`, and `SymmetryBoundary` (now deprecated).

File: mirgecom/boundary.py [hunk 1]
```diff
@@ -575,117 +596,202 @@ def __init__(self):
...
+
+    grad_t_bc = grad_temperature_bc_func(
+        dcoll=dcoll, dd_bdry=dd_bdry, gas_model=gas_model,
+        state_minus=state_minus, grad_cv_minus=grad_cv_minus,
+        grad_t_minus=grad_t_minus, **kwargs)
+
+    from mirgecom.viscous import viscous_flux
+    return viscous_flux(state_bc, grad_cv_bc, grad_t_bc) @ normal
+
+
```
Comment (MTCam) [inline mirgecom/boundary.py:730]: I had a refactor effort of the MengaldoBCs in #858 , which did something similar.  I merged our two efforts, and hope the result is acceptable to you.

File: mirgecom/boundary.py [hunk 2]
```diff
@@ -1370,114 +1446,85 @@ class IsothermalWallBoundary(PrescribedFluidBoundary):
...
+        normal = state_minus.array_context.thaw(dcoll.normal(dd_bdry))
 
-        internal_energy_plus = gas_model.eos.get_internal_energy(
-            temperature=temperature_wall, species_mass_fractions=mass_frac_plus)
+        mom_plus = self._no_slip.momentum_plus(state_minus.momentum_density, normal)
 
-        # Velocity is pinned to 0 here, no kinetic energy
-        total_energy_plus = state_minus.mass_density*internal_energy_plus
+        # Don't modify the energy, even though we modify the temperature; energy will
+        # be advected through the wall, which doesn't make sense
```
Comment (MTCam) [inline mirgecom/boundary.py:1490]: I hope that the flux, and any dissipation term vanishes in the energy component.  I will check the tests and add if necessary.

File: mirgecom/boundary.py [hunk 3]
```diff
@@ -1486,126 +1533,89 @@ class AdiabaticNoslipWallBoundary(PrescribedFluidBoundary):
...
 
-    def temperature_bc(self, state_minus, **kwargs):
-        """Get temperature value used in grad(T)."""
-        return state_minus.temperature
+        mom_bc = self._no_slip.momentum_bc(state_minus.momentum_density, normal)
 
-    def grad_cv_bc(self, state_minus, grad_cv_minus, normal, **kwargs):
+        # FIXME: Should we modify kinetic energy here? If not, add a comment
+        # explaining why
+        return replace_fluid_state(state_minus, gas_model, momentum=mom_bc)
```
Comment (MTCam) [inline mirgecom/boundary.py:1584]: Yeah, there's no K.E. here.  Should be fixed up.

File: mirgecom/boundary.py [hunk 4]
```diff
@@ -1486,126 +1533,89 @@ class AdiabaticNoslipWallBoundary(PrescribedFluidBoundary):
...
-        # *not* the numerical viscous flux as advised by [Bassi_1997]_.
-        f_ext = viscous_flux(state=state_plus, grad_cv=grad_cv_plus,
-                             grad_t=grad_t_plus)
+        return self._adiabatic.grad_temperature_bc(grad_t_minus, normal)
 
-        return f_ext@normal
-
-    def adiabatic_noslip_grad_av(self, grad_av_minus, **kwargs):
+    # FIXME: Remove this?
+    def grad_av_plus(self, grad_av_minus, **kwargs):
```
Comment (MTCam) [inline mirgecom/boundary.py:1606]: 👍

**Review by MTCam — COMMENTED** [review]
I really like these changes, and that the boundary interface is much more consistent now.  I merged this with a different refactor (#858) already and will push the result soon and ping you.

**Pull request merged**